silkrpc/core/estimate_gas_oracle.cpp

```diff
@@ -0,0 +1,131 @@
+#include <asio/post.hpp>
+#include <asio/use_awaitable.hpp>
+#include <silkrpc/core/blocks.hpp>
+#include <silkrpc/core/rawdb/chain.hpp>
+
+#include <silkrpc/common/log.hpp>
+
+namespace silkrpc::ego {
+
+asio::awaitable<intx::uint256> EstimateGasOracle::estimate_gas(const Call& call, uint64_t block_number) {
+    SILKRPC_DEBUG << "EstimateGasOracle::estimate_gas called\n";
+
+    std::uint64_t hi;
+    std::uint64_t lo = kTxGas - 1;
+
+    if (call.gas.value_or(0) >= kTxGas) {
+        SILKRPC_DEBUG << "Set HI with gas in args: " << call.gas.value_or(0) << "\n";
+        hi = call.gas.value();
+    } else {
+        // TODO(sixtysixter) worths it to test if returned header is not null?
+        const auto header = co_await block_header_provider_(block_number);
+        hi = header.gas_limit;
+        SILKRPC_DEBUG << "Evaluate HI with gas in block " << header.gas_limit << "\n";
+    }
+
+    intx::uint256 gas_price = call.gas_price.value_or(0);
+    if (gas_price != 0) {
+        evmc::address from = call.from.value_or(evmc::address{0});
+
+        std::optional<silkworm::Account> account{co_await account_reader_(from, block_number + 1)};
+
+        intx::uint256 balance = account->balance;
+        SILKRPC_DEBUG << "balance for address 0x" << from << ": 0x" << intx::hex(balance) << "\n";
+        if (call.value.value_or(0) > balance) {
+        // TODO(sixtysixter) what is the right code?
+            throw EstimateGasException{-1, "insufficient funds for transfer"};
+        }
+        auto available = balance - call.value.value_or(0);
+        int64_t allowance = int64_t(available / gas_price);
+        SILKRPC_DEBUG << "allowance: " << allowance << ", available: 0x" << intx::hex(available) << ", balance: 0x" << intx::hex(balance)  << "\n";
+        if (hi > allowance) {
+            SILKRPC_WARN << "gas estimation capped by limited funds: original " << hi
+                << ", balance 0x" << intx::hex(balance)
+                << ", sent" << intx::hex(call.value.value_or(0))
+                << ", gasprice" << intx::hex(gas_price)
+                << ", fundable" << allowance
+                << "\n";
+            hi = allowance;
+        }
+    }
+
+    if (hi > kGasCap) {
+        SILKRPC_WARN << "caller gas above allowance, capping: requested " << hi << ", cap " << kGasCap << "\n";
+        hi = kGasCap;
+    }
+    auto cap = hi;
+
+    SILKRPC_DEBUG << "hi: " << hi << ", lo: " << lo << ", cap: " << cap << "\n";
+
+    silkworm::Transaction transaction{call.to_transaction()};
+    while (lo + 1 < hi) {
+        auto mid = (hi + lo) / 2;
+        transaction.gas_limit = mid;
+
+        auto failed = co_await execution_test(transaction);
+
+        if (failed) {
+            lo = mid;
+        } else {
+            hi = mid;
+        }
+    }
+
+    if (hi == cap) {
+        SILKRPC_DEBUG << "HI == cap test again...\n";
+        transaction.gas_limit = hi;
+        auto failed = co_await execution_test(transaction);
```

Comment:
It seems that execution failure is not handled here

Reply:
Opps.... some line is missing!